In [3]:
import seaborn as sns
import pandas as pd

In [ ]:
mpg=sns.load_dataset('mpg')

In [ ]:
mpg.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [ ]:
# 3-c.
mpg_mean=pd.pivot_table(mpg, index='cylinders', values='mpg', aggfunc='mean')
mpg_mean

mpg
cylinders           
3          20.550000
4          29.286765
5          27.366667
6          19.985714
8          14.963107

In [ ]:
# 3-d.
horsepower_mean=pd.pivot_table(mpg, index='origin', values='hp_per_cc', aggfunc='mean')
horsepower_mean

In [ ]:
# 3-e.
mpg_model=pd.pivot_table(mpg[mpg['']])

ValueError: ignored

In [ ]:
mpg_model=pd.pivot_table(mpg, index='manufacturer', values='model', aggfunc='count').drop_duplicates()

In [ ]:
mpg_fmodel=mpg_model[mpg_model[mpg_model.values>=5]]

['0      chevrolet chevelle malibu\n1              buick skylark 320\n2             plymouth satellite\n3                  amc rebel sst\n4                    ford torino\n                 ...            \n393              ford mustang gl\n394                    vw pickup\n395                dodge rampage\n396                  ford ranger\n397                   chevy s-10\nName: name, Length: 398, dtype: object']


In [ ]:
mpg_fmodel_mean=mpg[[]]

'0      chevrolet chevelle malibu\n1              buick skylark 320\n2             plymouth satellite\n3                  amc rebel sst\n4                    ford torino\n                 ...            \n393              ford mustang gl\n394                    vw pickup\n395                dodge rampage\n396                  ford ranger\n397                   chevy s-10\nName: name, Length: 398, dtype: object'

# 타이타닉

In [ ]:
titanic = sns.load_dataset('titanic')

## a
타이타닉호의 승객에 대해 나이와 성별에 의한 카테고리 열인 category1 열을 만드시오. 
category1 카테고리는 다음과 같이 정의됨

1) 20살이 넘으면 성별을 그대로 사용한다.
2) 20살 미만이면 성별에 관계없이 “child”라고 한다.


In [ ]:
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [ ]:
titanic['category1'] =  titanic.apply(lambda x : "child" if x.age < 20 else "male" if x.sex=='male' else 'female', axis=1)
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone category1  
0      man        True  NaN  Southampton    no  False      male  
1    woman       False    C    Cherbourg   yes  False    female  
2    woman       False  NaN  Southampton   yes   True    female  
3    woman       False    C  Southampton   yes  False    female  
4      man        True  NaN  Southampton    no   True      male  
..     ...         ...  ...          ...   ...    ...       ...  
886    man        True  NaN  Southampton    no   True      male  
887  woman       False    B  Southampton   yes   True     child  
888  woman       False  NaN  Southampton    no  False    female  
889    man        True    C    Cherbourg   yes   True      male  
890    man        True  NaN   Queenstown    no   True      male  

[891 rows x 16 columns]

## b
 타이타닉호의 승객 중 나이를 명시하지 않은 고객은 나이를 명시한 고객의 평균 나이 값
이 되도록 titanic 데이터프레임을 고치시오.


In [ ]:
mean_age = titanic.age.mean(axis=0)
titanic.age.fillna(mean_age, inplace=True)

## c
성별, 선실(class)별, 출발지(embark_town)별 생존율을 구하시오.

In [ ]:
sex_group = titanic.groupby('sex')['survived']
sex_group.mean()

sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

In [ ]:
class_group = titanic.groupby('class')['survived']
class_group.mean()

class
First     0.629630
Second    0.472826
Third     0.242363
Name: survived, dtype: float64

In [ ]:
embark_town_group = titanic.groupby('embark_town')['survived']
embark_town_group.mean()

embark_town
Cherbourg      0.553571
Queenstown     0.389610
Southampton    0.336957
Name: survived, dtype: float64

## d
타이타닉호 승객을 ‘미성년자’, ‘청년’, ‘중년’, ‘장년’, ‘노년’ 나이 그룹으로 나누고, 각 그룹별 생존율을 구하시오.
* bins = [1, 20, 30, 50, 70, 100]
* labels = ["미성년자", "청년", "중년", "장년", "노년"]


In [ ]:
bins = [0, 20, 30, 50, 70, 100]
labels = ["미성년자", "청년", "중년", "장년", "노년"]
titanic["연령대"] = pd.cut(titanic.age, bins, labels=labels) 

titanic_age = pd.DataFrame(titanic.연령대.value_counts(), columns=["연령대"]) 
titanic_age["생존율"] = titanic.groupby('연령대')['survived'].mean()

In [ ]:
titanic_age

연령대       생존율
청년    407  0.334152
중년    241  0.423237
미성년자  179  0.458101
장년     59  0.355932
노년      5  0.200000

## e
qcut 명령으로 세 개의 나이 그룹을 만들고, 나이 그룹별 남녀 성비와 생존율을 구하시오.

In [ ]:
titanic

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone category1   연령대  
0     Third    man        True  NaN  Southampton    no  False      male    청년  
1     First  woman       False    C    Cherbourg   yes  False    female    중년  
2     Third  woman       False  NaN  Southampton   yes   True    female    청년  
3     First  woman       False    C  Southampton   yes  False    female    중년  
4     Third    man        True  NaN  Southampton    no   True      male    중년  
..      ...    ...         ...  ...          ...   ...    ...       ...   ...  
886  Second    man        True  NaN  Southampton    no   True      male    청년  
887   First  woman       False    B  Southampton   yes   True     child  미성년자  
888   Third  woman       False  NaN  Southampton    no  False    female    청년  
889   First    man        True    C    Cherbourg   yes   True      male    청년  
890   Third    man        True  NaN   Queenstown    no   True      male    중년  

[891 rows x 17 columns]

In [ ]:
titanic['age-group'] = pd.qcut(titanic.age, 3, labels=['젊은이', '중간', '늙은이'])
titanic

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone category1   연령대  \
0     Third    man        True  NaN  Southampton    no  False      male    청년   
1     First  woman       False    C    Cherbourg   yes  False    female    중년   
2     Third  woman       False  NaN  Southampton   yes   True    female    청년   
3     First  woman       False    C  Southampton   yes  False    female    중년   
4     Third    man        True  NaN  Southampton    no   True      male    중년   
..      ...    ...         ...  ...          ...   ...    ...       ...   ...   
886  Second    man        True  NaN  Southampton    no   True      male    청년   
887   First  woman       False    B  Southampton   yes   True     child  미성년자   
888   Third  woman       False  NaN  Southampton    no  False    female    청년   
889   First    man        True    C    Cherbourg   yes   True      male    청년   
890   Third    man        True  NaN   Queenstown    no   True      male    중년   

    age-group  
0         젊은이  
1         늙은이  
2          중간  
3         늙은이  
4         늙은이  
..        ...  
886        중간  
887       젊은이  
888        중간  
889        중간  
890       늙은이  

[891 rows x 18 columns]

In [ ]:
age_group_survived = titanic.groupby('age-group')['survived']
age_group_survived.mean()

age-group
젊은이    0.411960
중간     0.335526
늙은이    0.405594
Name: survived, dtype: float64

In [ ]:
titanic['sex_category'] = titanic.apply(lambda x : 1 if x.sex=='male' else 0, axis=1)
titanic['sex_category_f'] = titanic.apply(lambda x : 1 if x.sex=='female' else 0, axis=1)

titanic

survived  pclass     sex        age  sibsp  parch     fare embarked  \
0           0       3    male  22.000000      1      0   7.2500        S   
1           1       1  female  38.000000      1      0  71.2833        C   
2           1       3  female  26.000000      0      0   7.9250        S   
3           1       1  female  35.000000      1      0  53.1000        S   
4           0       3    male  35.000000      0      0   8.0500        S   
..        ...     ...     ...        ...    ...    ...      ...      ...   
886         0       2    male  27.000000      0      0  13.0000        S   
887         1       1  female  19.000000      0      0  30.0000        S   
888         0       3  female  29.699118      1      2  23.4500        S   
889         1       1    male  26.000000      0      0  30.0000        C   
890         0       3    male  32.000000      0      0   7.7500        Q   

      class    who  adult_male deck  embark_town alive  alone category1   연령대  \
0     Third    man        True  NaN  Southampton    no  False      male    청년   
1     First  woman       False    C    Cherbourg   yes  False    female    중년   
2     Third  woman       False  NaN  Southampton   yes   True    female    청년   
3     First  woman       False    C  Southampton   yes  False    female    중년   
4     Third    man        True  NaN  Southampton    no   True      male    중년   
..      ...    ...         ...  ...          ...   ...    ...       ...   ...   
886  Second    man        True  NaN  Southampton    no   True      male    청년   
887   First  woman       False    B  Southampton   yes   True     child  미성년자   
888   Third  woman       False  NaN  Southampton    no  False    female    청년   
889   First    man        True    C    Cherbourg   yes   True      male    청년   
890   Third    man        True  NaN   Queenstown    no   True      male    중년   

    age-group  sex_category  sex_category_f  
0         젊은이             1               0  
1         늙은이             0               1  
2          중간             0               1  
3         늙은이             0               1  
4         늙은이             1               0  
..        ...           ...             ...  
886        중간             1               0  
887       젊은이             0               1  
888        중간             0               1  
889        중간             1               0  
890       늙은이             1               0  

[891 rows x 20 columns]

In [ ]:
# 나이 그룹별 남자비율
age_group_survived = titanic.groupby('age-group')['sex_category']
age_group_survived.mean()

age-group
젊은이    0.594684
중간     0.680921
늙은이    0.667832
Name: sex_category, dtype: float64

In [ ]:
# 나이 그룹별 여자비율
age_group_survived_f = titanic.groupby('age-group')['sex_category_f']
age_group_survived_f.mean()

age-group
젊은이    0.405316
중간     0.319079
늙은이    0.332168
Name: sex_category_f, dtype: float64

# Iris

In [4]:
iris = sns.load_dataset('iris')
iris

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

## a
붓꽃 종(species)별로 꽃잎길이(sepal_length), 꽃잎폭(sepal_width),
꽃받침길이(petal_length), 꽃받침폭(petal_width)의 평균, 표준편차 등
기초통계량(describe())을 구하시오.

In [12]:
iris.groupby('species').describe()

sepal_length                                              \
                  count   mean       std  min    25%  50%  75%  max   
species                                                               
setosa             50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8   
versicolor         50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0   
virginica          50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9   

           sepal_width         ... petal_length      petal_width         \
                 count   mean  ...          75%  max       count   mean   
species                        ...                                        
setosa            50.0  3.428  ...        1.575  1.9        50.0  0.246   
versicolor        50.0  2.770  ...        4.600  5.1        50.0  1.326   
virginica         50.0  2.974  ...        5.875  6.9        50.0  2.026   

                                               
                 std  min  25%  50%  75%  max  
species                                        
setosa      0.105386  0.1  0.2  0.2  0.3  0.6  
versicolor  0.197753  1.0  1.2  1.3  1.5  1.8  
virginica   0.274650  1.4  1.8  2.0  2.3  2.5  

[3 rows x 32 columns]

## b
3분위수(Q3)와 1분위수(Q1)의 차이보다 1.5배가 크거나 작은 데이터는 이상치이다.
즉,
* Q1 - 1.5 * (Q3 - Q1) 보다 작은 데이터
* Q3 + 1.5 * (Q3 - Q1) 보다 큰 데이터

이 이상치를 제거하고 위의 4가지 항목에 대해서 평균, 표준편차를 구하시오.